### PR TITLE
App static layout structure (plus...)

### PR DESCRIPTION
**Pull request by shon-button**
Addresses [ticket 175](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/175)

## 🚧 Changes

✨ Components:
    - Add client\middleware.ts for request routing management
    - Add client\middlewares\ * files for task specific functions applied to a request (TBD)
    - Add client\app\auth\signin\page.tsx for login route (TBD)
    - Add client\app\components\theme\* for MUI custom theme with Next.js 13 app router
    - Modify client\app\layout.tsx to use ThemeRegistry
    - Modify client\next.config.js for MUI imports
    - Add client\app\components\layout\Header for static app layout
    - Add client\app\components\layout\Footer for static app layout
    
🧹 Cleanup:
    - Remove Tailwindcss configurations files
    - Modify package.json to resolve dependancy warnings
    - Remove client\app\styles\global.css
    - Remove client\app\styles\layout.module.css    
    - Rename client\app\utils\actions\operationSubmitHandler.ts to utils\actions.ts, as per Next.js convention
    - Add client\.env\API_URL variable for API requests URL
    - Add client\app\utils\api.ts fetch utility function to handle API requests and error catching
    - Re-code fetch calls to hardcoded API to use utils\api.ts fetch utility function, as per DRY convention
    - remove client\app\components\loading.tsx
    
🏗️ Redesign:
    - Delete client\app\components\loading.tsx for suspend loading effect
    - Add client\app\components\loading\SkeletonGrid.tsx for suspend loading effect
    - Add client\app\components\loading\SkeletonForm.tsx for suspend loading effect
    - Add client\app\components\routes\operations\Operations.tsx (contains async components from client\app\operations\page.tsx)
    - Recode client\app\operations\page.tsx to wrap asynchronous client\app\components\operations\Operations.tsx using <Suspense>
    - Add client\app\components\routes\operations\form\Operation.tsx (contains async components from client\app\operations\create\page.tsx)
    - Recode client\app\operations\create\page.tsx to wrap asynchronous client\app\components\operations\form\Operation.tsx using <Suspense> 
    

 ## 🚀 Impact:
   - Material UI configured with the Next.js App Router
   - Initial static layout structure (**rudimentary and non-functional- refinement TBD**)
   -  Middleware request routing architecture in place for future requirements of request rewriting, redirecting, or modifying before a response
   - Re-architecture of app router files for wrapping asynchronous components in page routes to improved user interaction. 
**Ex:** 
Route http://localhost:3000/operations displays page.tsx which displays none dynamic elements and uses <Suspense> with a loading skeleton to indicate waiting for the async data components

## 📝  Notes:
- Added a temp latency simulation on async data to emphasize the user interaction with <Suspense> skeleton loading

##  🔬 Testing:

1. Open terminal in cas-registration\bc_obps directory
2. Follow the [developer environment setup](https://github.com/bcgov/cas-registration/blob/develop/docs/developer-environment-setup.md#backend-environment-setup) to start the API
3. Open a terminal in cas-registration\client folder
4. Install changes to the packgae.json
   ```
   yarn
   ```
5. Start the development server:
   ```
   yarn dev
   ```
6. In your web browser, navigate to `http://localhost:3000/`
7. You should see "navigation" page with MUI header and footer
image
8. Click link "Operations"
9. You should see "Operations List" page with loading skeleton
image
10. Once loading has completed you should see list of operations
11. Click button "Add Operation"
12. You should see "Create a New Operation" page with loading skeleton
image
13. Once loading is complete you should see an empty form
14. Complete and submit form
15. A success message should display
16. Click the back button
17. New record should display
18. Click the "View Details" on a row
19.. You should see "Edit an Operation" page with loading skeleton
image
20.  Once loading is complete you should see an form with data from the selected row.
21. Edit submit form
22. A success message should display
23. Click the back button
24. Edits should display

### CODE REVIEW

**Review by BCerki — COMMENTED** [review]
This looks amazing! I have a couple general comments:
- The submit button is cut off by the footer on my screen: 
Screenshot from 2023-10-06 13-07-00
- I think we could use some unit tests for the new layout components (operations pages and forms will be handled in #139). And for MUI theme stuff if that's useful? Here are examples of the kind of layout stuff we tested in CIF: https://github.com/bcgov/cas-cif/tree/develop/app/tests/unit/components/Layout. Unit tests are probably a separate PR thing though since we haven't got jest playing nice with jsx yet. 
- I get a left sidebar on the signout: 
image

File: client/middleware.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import { stackMiddlewares } from "@/middlewares/stackMiddlewares";
+import { withAuthorization } from "@/middlewares/withAuthorization";
+import { withResponse } from "@/middlewares/withResponse";
+
+/* 📌
```
[the remark this answers is not in the document]
Comment (shon-button) [reply client/middleware.ts:5]: Hi Brianna, thank you so much for the detailed review!
If it is ok with you, I would like to capture your additions/suggestions in a separate ticket specific to refinements needed to the current rudimentary layout so to be able to merge the most relevant parts of this PR:
- Material UI and Next.js App Router architecture 
- Middleware request routing architecture 
- Wrapping asynchronous components with <Suspense> architecture

File: client/package.json [hunk 2]
```diff
@@ -33,13 +33,13 @@
...
     "eslint-config-airbnb-typescript": "^17.1.0",
+    "eslint-plugin-import": "^2.25.3",
     "eslint-config-next": "13.4.19",
     "next": "^13.5.4",
     "postcss": "^8.4.30",
     "react": "18.2.0",
     "react-dom": "18.2.0",
     "react-jsonschema-form": "^1.8.1",
-    "tailwindcss": "^3.3.3",
     "ts-node": "^10.9.1",
```
[the remark this answers is not in the document]
Comment (shon-button) [reply client/package.json:43]: Agree. Seems we can integrate any other  CSS library...with this caveat:

CSS injection order

By default, Emotion injects Material UI styles at the bottom of the HTML <head>, which gives them precedence over custom styles—for example if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS.

Emotion provides the prepend: true option for createCache to reverse the injection order, so custom styles can override Material UI styles without using !important.

Currently, prepend does not work reliably with the App Router, but you can work around it by wrapping Emotion styles in a CSS @layer with a modification to the snippet above:

File: client/middleware.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (BCerki) [reply client/middleware.ts:5]: Sounds good! I'll approve once the checks are passing

**Comment by shon-button** [comment]
Deleted auth demo folder with error.jsx page as SonarCloud Code Analysis failed with: 

Do not use "Error" to declare a function - use another name.
export default function Error() {


Need to determine how to handle [Netx.js error.jsx](https://nextjs.org/docs/app/building-your-application/routing/error-handling) with SonarCloud Code Analysis

**Comment by BCerki** [comment]
Also, just thought of it, are you planning to take out the delay?

**Review by BCerki — APPROVED** [review]

**Pull request merged**